scripts/exact_1d_bayesian_filtering_utils.py

```diff
@@ -0,0 +1,384 @@
+from typing import Callable
+import functools
+
+import jax
+from jax import numpy as jnp
+from jax import scipy as jsp
+from jax import random
+import matplotlib.pyplot as plt
+
+
+plt.rcParams.update({"font.size": 16})
+
+
+def generate_x_true(rng_key, K, x0_rvs: Callable, v_rvs: Callable, f: Callable):
+    rng_keys = random.split(rng_key, num=2)
+    x0 = x0_rvs(rng_keys[0], shape=())
+    x_true = jnp.zeros(shape=(K + 1,))
+    x_true = x_true.at[0].set(x0)
+    v = v_rvs(rng_keys[1], shape=x_true.shape)
+    for k in range(1, K + 1):
+        x_true = x_true.at[k].set(f(x_true[k - 1], v[k - 1], k=k - 1))
+    return jnp.array(x_true)
+
+
+def generate_y(rng_key, x_true, e_rvs: Callable, h: Callable):
+    shape = x_true.shape
+    e = e_rvs(rng_key, shape=shape)
+    y = h(x_true, e)
+    y = y.at[0].set(jnp.inf)
+    return y
+
+
+def x_pdf(x_new, x, k, v_pdf, f):
+    v = x_new - f(x=x, v=0, k=k)
+    return v_pdf(x - v)
+
+
+def y_likelihood(y, x, e_pdf, h):
+    e = h(x=x, e=0)
+    return e_pdf(y - e)
+
+
+def inversion_sampling(rng_key, x_grid, px_grid, num_samples):
+    rng_keys = random.split(rng_key, num=2)
+    u = random.uniform(rng_keys[0], shape=(num_samples, 1))
+    delta = x_grid[1] - x_grid[0]
+    noise = random.uniform(
+        rng_keys[1], minval=-delta / 2, maxval=delta / 2, shape=(num_samples,)
+    )
+    # It only works for sufficient dense uniformly spaced grid
+    point_mass = px_grid
+
+    cdf = jnp.cumsum(point_mass)
+
+    bound_cdf = jnp.where(cdf < u, cdf, 0)
+    idx = jnp.argmax(bound_cdf, axis=1)
+    x = x_grid[idx]
+    return x + noise
+
+
+def kde(x_grid, x, kernel_variance):
+    delta = x_grid[1] - x_grid[0]
+
+    px = jax.vmap(functools.partial(jsp.stats.norm.pdf, loc=x, scale=kernel_variance))(
+        x_grid
+    )
+    px = jnp.sum(px, axis=1)
+    px = px / jnp.sum(px) / delta
+    return px
+
+
+def novel_density(
+    rng_key,
+    y,
+    x_grid,
+    x0_pdf,
+    v_rvs: Callable,
+    e_rvs: Callable,
+    f: Callable,
+    h: Callable,
+    num_samples: int,
+    K,
+    kernel_variance,
```

Comment:
Don't break the lines of the function after defining every parameter of the function.

Reply:
Hi @gerdm, Do u mean that I should write the code like this?
```
def novel_density(
    rng_key, y, x_grid,
    x0_pdf, v_rvs: Callable, e_rvs: Callable,
    f: Callable, h: Callable, num_samples: int,
    K, kernel_variance,
):
```

Reply:
That looks much better! I would just modify the last line to be pep-8 compliant:

```
def novel_density(
    rng_key, y, x_grid,
    x0_pdf, v_rvs: Callable, e_rvs: Callable,
    f: Callable, h: Callable, num_samples: int,
    K, kernel_variance):
```

I would also suggest that if you're using type hints, be sure to use them consistently throughout the definitions of your functions.